pkg/apis/core/types.go

```diff
@@ -2246,6 +2273,10 @@ type Container struct {
 	// Compute resource requirements.
 	// +optional
 	Resources ResourceRequirements
+	// Resources resize policy for the container.
+	// +featureGate=InPlacePodVerticalScaling
+	// +optional
+	ResizePolicy []ContainerResizePolicy
```

Comment:
The default for a list is zero elements.  You spec the default value for `ContainerResizePolicy.Policy` but you don't spec a default for this field.  What happens if the user does not set this?  I presume it's something like

"""
Any resource which is specified in `resources` which is not specified in `resizePolicy` is assumed to be RestartNotRequired.
"""

The disadvantages of specifying it that way are:
1) all consumer code needs to handle this case and hardcode the default
2) you can't change the default (e.g. in v2)

A different way to spec this would be to ensure that this list has exactly the same set of resources as `resources` and set the default value in there.  If someone updates resources to remove a key, you remove it here, and if they add a key, add it here.

That's a little ugly because it modifies user input, but we can't tell the difference between "the use set it" and "we set it for them", and it's horrible to ask them to clear fields they didn't set in the first place.  Lesser evil, IMO.

Reply:
> The disadvantages of specifying it that way are:
> 
> 1. all consumer code needs to handle this case and hardcode the default
> 2. you can't change the default (e.g. in v2)
> 
> A different way to spec this would be to ensure that this list has exactly the same set of resources as `resources` and set the default value in there. If someone updates resources to remove a key, you remove it here, and if they add a key, add it here.
> 

Yes what you described is the behavior I intended. resizePolicy is immutable. (We expect user sets this based on knowledge of their app requirements, and that does not change mid-flight) The user does not have to set anything if they just want the defaults (RestartNotRequired) or see any defaults they didn't set. 

Can you please help me understand point 2? If we changed default from RestartNotRequired to RestartRequired, any user yaml that doesn't set it will get the new behavior. Any that sets it will continue to get the behavior they asked for. What am I missing? Are you referring to yaml with apiVersion v1 vs v2?

Reply:
There are 2 kinds of defaults: the "if this field is empty, clients should ..." kind and the "if you don't specify this field, we will fill it in with ..." kind.  Kubernetes almost always uses the second kind because it is explicit and visible.  And it has a nice side-effect that you can change the defaults with API versions.

Consider the first kind (leave it empty).  User submits a pod, we leave this field empty, kubelet inteprets empty to mean "NotRequired".  Over time we decide the default SHOULD BE "NotAllowed" (hypothetical, obv.).  We can never change it because that would be a breaking change for anyone who didn't specify it.

Now consider the second kind.  User User submits a pod, we see this field is empty, so we explicitly fill it in with values.  Kubelet sees those values and acts on them.  Over time we decide the default SHOULD BE "NotAllowed" and when we eventually make a core/v2 API, we change the default.  Pods which come in on v1 get the old default, pods which come in on v2 get the new.

To do that, you would have to add a block of code that says:

```
for each key in resources.{requests,limits} {
    if that key does not exist in resizePolicy {
        add it with the default value
    }
}
```

If we want to support updating the keys (e.g. adding/removing) it gets a bit more complicated, I think, but I'd have to explore it.

I, obviously, prefer explicit

Reply:
Understood, it is fair. I'll look into setting it up in defaults.go. From what I see, it should not require any changes to current kubelet code and should not trigger a reset Derek's review. If it does, we can add it in a separate PR after merge (I'll test it before merge)

Reply:
thanks.  I will accept a followup if you want :)

Reply:
The change itself is small and works great on initial bunch of manual tests. But it also needs various unit test fixes and it is some work to fix and review. I'll do it as a follow-up PR.

I have added a change that fixes comments only: https://github.com/kubernetes/kubernetes/pull/102884/commits/a9ed85fc56b994039397a76dea64415e24440c44
Please do a quick scan.